File: 1035_uncrossed_lines/max_uncrossed_lines.go
package main

/**
 * Time: O(m * n)
 * Space: O(m * n)
 */
func maxUncrossedLines(A []int, B []int) int {
	// bottom-up approach
	m, n := len(A), len(B)
	dp := make([][]int, m+1)
	for i := 0; i < m+1; i++ {
		dp[i] = make([]int, n+1)
	}
	for r := m - 1; r >= 0; r-- {
		for c := n - 1; c >= 0; c-- {
			if A[r] == B[c] {
				dp[r][c] = dp[r+1][c+1] + 1
			} else {
				dp[r][c] = max(dp[r][c+1], dp[r+1][c])
			}
		}
	}
	return dp[0][0]
}

func max(x, y int) int {
	if x >= y {
		return x
	}
	return y
}

func main() {

}

File: 1035_uncrossed_lines/max_uncrossed_lines.py
from typing import List


def max_uncrossed_lines(A: List[int], B: List[int]) -> int:
    """
    Time: O(n * m)
    Space: O(n * m)
    Bottom up approach.
    """
    A = [0] + A
    B = [0] + B
    m, n = len(A), len(B)
    dp = [[0] * n for _ in range(m)]
    for i in range(1, m):
        for j in range(1, n):
            if A[i] == B[j]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
    return dp[-1][-1]
